File: src/main/java/br/com/jiratorio/mapper/EstimateIssueMapper.java
package br.com.jiratorio.mapper;

import br.com.jiratorio.domain.EstimateIssue;
import br.com.jiratorio.domain.FluxColumn;
import br.com.jiratorio.domain.changelog.JiraChangelog;
import br.com.jiratorio.domain.changelog.JiraChangelogHistory;
import br.com.jiratorio.domain.changelog.JiraChangelogItem;
import br.com.jiratorio.domain.entity.Board;
import br.com.jiratorio.domain.entity.embedded.Changelog;
import br.com.jiratorio.service.HolidayService;
import br.com.jiratorio.util.DateUtil;
import br.com.jiratorio.util.ParseUtil;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.Collection;
import java.util.Iterator;
import java.util.List;
import java.util.Objects;
import java.util.Set;
import java.util.stream.Collectors;
import java.util.stream.StreamSupport;
import lombok.SneakyThrows;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Component;
import org.springframework.util.StringUtils;

@Slf4j
@Component
public class EstimateIssueMapper {

    private final HolidayService holidayService;

    private final ObjectMapper objectMapper;

    public EstimateIssueMapper(final HolidayService holidayService,
                               final ObjectMapper objectMapper) {
        this.holidayService = holidayService;
        this.objectMapper = objectMapper;
    }

    @SneakyThrows
    public List<EstimateIssue> parseEstimate(final String rawText, final Board board) {
        JsonNode jsonNode = objectMapper.readTree(rawText);
        Iterator<JsonNode> issues = jsonNode.path("issues").elements();

        List<LocalDate> holidays = holidayService.findDaysByBoard(board.getId());

        FluxColumn fluxColumn = new FluxColumn(board);
        Set<String> startColumns = fluxColumn.getStartColumns();

        Iterable<JsonNode> iterable = () -> issues;
        return StreamSupport.stream(iterable.spliterator(), false)
                .map(issue -> {
                    JsonNode fields = issue.path("fields");

                    List<JiraChangelogItem> changelogItems = extractChangelogItems(issue);
                    List<Changelog> changelog = ParseUtil.parseChangelog(changelogItems, holidays, board.getIgnoreWeekend());
                    if (!changelog.isEmpty()) {
                        Changelog changelogItem = changelog.get(changelog.size() - 1);
                        changelogItem.setLeadTime(DateUtil.daysDiff(changelogItem.getCreated(), LocalDateTime.now(),
                                holidays, board.getIgnoreWeekend()));
                        changelogItem.setEndDate(LocalDateTime.now());
                    }
                    LocalDateTime created = DateUtil.parseFromJira(fields.get("created").asText());

                    LocalDateTime startDate = null;

                    for (Changelog cl : changelog) {
                        if (startDate == null && startColumns.contains(cl.getTo())) {
                            startDate = cl.getCreated();
                        }
                    }
                    if ("BACKLOG".equals(board.getStartColumn())) {
                        startDate = created;
                    }
                    if (startDate == null) {
                        return null;
                    }
                    String priority = null;
                    if (fields.has("priority") && !fields.get("priority").isNull() && !fields.path("priority").isMissingNode()) {
                        priority = fields.path("priority").get("name").asText(null);
                    }

                    Long leadTime = DateUtil.daysDiff(startDate, LocalDateTime.now(), holidays, board.getIgnoreWeekend());

                    Long timeInImpediment = ParseUtil.countTimeInImpediment(board, changelogItems, changelog, LocalDateTime.now(), holidays);

                    String author = null;
                    JsonNode creator = fields.path("creator");
                    if (creator != null && !creator.isNull()) {
                        author = creator.get("displayName").asText(null);
                    }
                    return EstimateIssue.builder()
                            .creator(author)
                            .key(issue.get("key").asText(null))
                            .issueType(fields.path("issuetype").get("name").asText(null))
                            .created(created)
                            .startDate(startDate)
                            .leadTime(leadTime)
                            .system(parseElement(fields, board.getSystemCF()))
                            .epic(parseElement(fields, board.getEpicCF()))
                            .estimated(parseElement(fields, board.getEstimateCF()))
                            .project(parseElement(fields, board.getProjectCF()))
                            .summary(fields.get("summary").asText())
                            .changelog(changelog)
                            .impedimentTime(timeInImpediment)
                            .priority(priority)
                            .build();
                })
                .filter(Objects::nonNull)
                .collect(Collectors.toList());
    }

    @SneakyThrows
    private List<JiraChangelogItem> extractChangelogItems(final JsonNode issue) {
        JiraChangelog changelog = objectMapper.readValue(issue.path("changelog").toString(), JiraChangelog.class);
        changelog.getHistories().forEach(cl -> cl.getItems().forEach(i -> i.setCreated(cl.getCreated())));
        return changelog.getHistories().parallelStream()
                .map(JiraChangelogHistory::getItems)
                .flatMap(Collection::stream)
                .collect(Collectors.toList());
    }

    private String parseElement(final JsonNode fields, final String cf) {
        if (StringUtils.isEmpty(cf)) {
            return null;
        }
        final JsonNode jsonElement = fields.path(cf);
        if (jsonElement == null || jsonElement.isMissingNode()) {
            return null;
        }

        if (jsonElement.isArray()) {
            Iterator<JsonNode> elements = jsonElement.elements();
            if (elements == null || !elements.hasNext()) {
                return null;
            }
            Iterable<JsonNode> iterable = () -> elements;
            return StreamSupport.stream(iterable.spliterator(), true)
                    .map(component -> component.isContainerNode()
                            ? component.get("name").asText(null)
                            : component.asText(null))
                    .findFirst().orElse(null);
        }

        if (jsonElement.isContainerNode()) {
            return jsonElement.get("value").asText(null);
        }

        return jsonElement.asText(null);
    }

}

File: src/main/java/br/com/jiratorio/mapper/EstimateIssueMapper.kt
package br.com.jiratorio.mapper

import br.com.jiratorio.domain.FluxColumn
import br.com.jiratorio.domain.changelog.JiraChangelog
import br.com.jiratorio.domain.changelog.JiraChangelogItem
import br.com.jiratorio.domain.entity.Board
import br.com.jiratorio.domain.estimate.EstimateIssue
import br.com.jiratorio.extension.extractValue
import br.com.jiratorio.service.ChangelogService
import br.com.jiratorio.service.HolidayService
import br.com.jiratorio.util.DateUtil
import com.fasterxml.jackson.databind.JsonNode
import com.fasterxml.jackson.databind.ObjectMapper
import org.springframework.stereotype.Component
import java.time.LocalDate
import java.time.LocalDateTime

@Component
class EstimateIssueMapper(
    private val holidayService: HolidayService,
    private val objectMapper: ObjectMapper,
    private val changelogService: ChangelogService
) {

    fun parseEstimate(rawText: String, board: Board): List<EstimateIssue> {
        val holidays = holidayService.findDaysByBoard(board.id)

        val fluxColumn = FluxColumn(board)
        val startColumns = fluxColumn.startColumns

        return objectMapper.readTree(rawText).path("issues").mapNotNull {
            parseIssue(it, board, startColumns, holidays)
        }
    }

    fun parseIssue(
        issue: JsonNode,
        board: Board,
        startColumns: Set<String>,
        holidays: List<LocalDate>
    ): EstimateIssue? {
        val fields = issue.path("fields")

        val changelogItems = extractChangelogItems(issue)
        val changelog = changelogService.parseChangelog(changelogItems, holidays, board.ignoreWeekend)
        if (!changelog.isEmpty()) {
            val changelogItem = changelog[changelog.size - 1]
            changelogItem.leadTime =
                DateUtil.daysDiff(
                    changelogItem.created!!,
                    LocalDateTime.now(),
                    holidays,
                    board.ignoreWeekend
                )
            changelogItem.endDate = LocalDateTime.now()
        }
        val created = DateUtil.parseFromJira(fields.get("created").asText())

        var startDate: LocalDateTime? = null

        for (cl in changelog) {
            if (startDate == null && startColumns.contains(cl.to)) {
                startDate = cl.created
            }
        }

        if ("BACKLOG" == board.startColumn) {
            startDate = created
        }
        if (startDate == null) {
            return null
        }

        val priority: String? =
            if (fields.hasNonNull("priority")) {
                fields.path("priority").extractValue()
            } else {
                null
            }

        val leadTime = DateUtil.daysDiff(startDate, LocalDateTime.now(), holidays, board.ignoreWeekend)

        val timeInImpediment = board.impedimentType?.timeInImpediment(
            board.impedimentColumns,
            changelogItems,
            changelog,
            LocalDateTime.now(),
            holidays,
            board.ignoreWeekend
        ) ?: 0

        val author: String? =
            if (fields.hasNonNull("creator")) {
                fields.path("creator").path("displayName").extractValue()
            } else {
                null
            }

        return EstimateIssue(
            creator = author,
            key = issue.get("key").extractValue(),
            issueType = fields.path("issuetype").extractValue(),
            created = created,
            startDate = startDate,
            leadTime = leadTime,
            system = fields.path(board.systemCF).extractValue(),
            epic = fields.path(board.epicCF).extractValue(),
            estimated = fields.path(board.estimateCF).extractValue(),
            project = fields.path(board.projectCF).extractValue(),
            summary = fields.get("summary").extractValue(),
            changelog = changelog,
            impedimentTime = timeInImpediment,
            priority = priority
        )
    }

    private fun extractChangelogItems(issue: JsonNode): List<JiraChangelogItem> {
        val changelog = objectMapper.treeToValue(issue.path("changelog"), JiraChangelog::class.java)
        changelog.histories.forEach { cl -> cl.items.forEach { i -> i.created = cl.created } }
        return changelog.histories.map { it.items }.flatten()
    }

}
